partners_api/uber_api.cpp

```diff
@@ -0,0 +1,221 @@
+}
+}  // namespace
+
+namespace uber
+{
+// static
+string RawApi::GetProducts(ms::LatLon const & pos)
+{
+  stringstream url;
+  url << "https://api.uber.com/v1/products?server_token=" << UBER_SERVER_TOKEN
+      << "&latitude=" << pos.lat << "&longitude=" << pos.lon;
+
+  return RunSimpleHttpRequest(url.str());
+}
+
+// static
+string RawApi::GetEstimatedTime(ms::LatLon const & pos)
+{
+  stringstream url;
+  url << "https://api.uber.com/v1/estimates/time?server_token=" << UBER_SERVER_TOKEN
+      << "&start_latitude=" << pos.lat << "&start_longitude=" << pos.lon;
+
+  return RunSimpleHttpRequest(url.str());
+}
+
+// static
+string RawApi::GetEstimatedPrice(ms::LatLon const & from, ms::LatLon const & to)
+{
+  stringstream url;
+  url << "https://api.uber.com/v1/estimates/price?server_token=" << UBER_SERVER_TOKEN
+      << "&start_latitude=" << from.lat << "&start_longitude=" << from.lon
+      << "&end_latitude=" << to.lat << "&end_longitude=" << to.lon;
+
+  return RunSimpleHttpRequest(url.str());
+}
+
+void ProductMaker::Reset(size_t const requestId)
+{
+  lock_guard<mutex> lock(m_mutex);
+
+  m_requestId = requestId;
+  m_times.reset();
+  m_prices.reset();
+}
+
+void ProductMaker::SetTimes(size_t const requestId, string const & times)
+{
+  lock_guard<mutex> lock(m_mutex);
+
+  if (requestId != m_requestId)
+    return;
+
+  m_times = make_unique<string>(times);
+}
+
+void ProductMaker::SetPrices(size_t const requestId, string const & prices)
+{
+  lock_guard<mutex> lock(m_mutex);
+
+  if (requestId != m_requestId)
+    return;
+
+  m_prices = make_unique<string>(prices);
+}
+
+void ProductMaker::MakeProducts(size_t const requestId, ProductsCallback const & fn)
+{
+  lock_guard<mutex> lock(m_mutex);
+
+  if (requestId != m_requestId || !m_times || !m_prices)
+    return;
+
+  vector<uber::Product> products;
+
+  if (m_times->empty() || m_prices->empty())
+  {
+    LOG(LWARNING, ("Time or price is empty, time:", *m_times, "; price:", *m_prices));
+    fn(products, m_requestId);
+    return;
+  }
+
+  try
+  {
+    my::Json timesRoot(m_times->c_str());
+    my::Json pricesRoot(m_prices->c_str());
+    auto const timesArray = json_object_get(timesRoot.get(), "times");
+    auto const pricesArray = json_object_get(pricesRoot.get(), "prices");
+    if (CheckUberAnswer(timesArray) && CheckUberAnswer(pricesArray))
+    {
+      FillProducts(timesArray, pricesArray, products);
+    }
+  }
+  catch (my::Json::Exception const & e)
+  {
+    LOG(LERROR, (e.Msg()));
+    products.clear();
+  }
+
+  fn(products, requestId);
+}
+
+size_t Api::GetAvailableProducts(ms::LatLon const & from, ms::LatLon const & to,
+                                 ProductsCallback const & fn)
+{
+  static size_t requestId = 0;
+  size_t reqId = ++requestId;
+
+  m_maker->Reset(reqId);
+
+  threads::SimpleThread([this, from, reqId, fn]()
+  {
+    m_maker->SetTimes(reqId, uber::RawApi::GetEstimatedTime(from));
+    m_maker->MakeProducts(reqId, fn);
```

Comment:
What's the point to call MakeProducts here? Why not to do following in SetTimes and SetPrices:

```c++
void SetTimes(...)
{
   ...
   MakeProducts(reqId);
}

void SetPrices(...)
{
  ...
  MakeProducts(reqId);
}
```
and in make products all what you need is to check that both fields (time and prices) are ready, and then call a callback.